### DISCUSSION TITLE
Uniq function optimize table for better performance

### DISCUSSION THREAD

**Discussion by vigneshshettyin** [ClickHouse/ClickHouse#57862] [Q&A]
```
create table final_table (
  CMO_ID String, 
  ROLL_UP_DATE Date, 
  CODE String, 
  CLP_ID Int64,
  PLP_ID Int64
)
ENGINE = MergeTree
PARTITION BY toYYYYMM(ROLL_UP_DATE)
ORDER BY (CODE, CMO_ID, ROLL_UP_DATE)
SETTINGS index_granularity = 8192
```

On top of this table I have created a Distributed table , Sharding key CMO_ID - CityHash

```
SELECT
    CMO_ID,
    uniq(PLP_ID) AS PLP,
    uniq(CLP_ID) AS CLP
FROM distributed_final_table
WHERE CODE IN ('LOP1', 'LOP2', 'LOP3', 'LOP4', 'LOP5', 'LOP6', 'LOP7', 'LOP8', 'LOP9', 'LOP10', 'LOP11', 'LOP12', 'LOP13', 'LOP14', 'LOP15', 'LOP16', 'LOP17', 'LOP18', 'LOP19', 'LOP20', 'LOP21', 'LOP22')
GROUP BY CMO_ID
LIMIT 10
```

-- Above query does full table scan as I have included all CODES for full date range (Date range by Month)

Query timeout due to memory error

```
Received exception from server (version 23.11.1):
Code: 241. DB::Exception: Received from localhost:9000. DB::Exception: Memory limit (total) exceeded: would use 27.69 GiB (attempt to allocate chunk of 0 bytes), maximum: 27.69 GiB. OvercommitTracker decision: Query was selected to stop by OvercommitTracker.: While executing AggregatingTransform. (MEMORY_LIMIT_EXCEEDED)
```

I have also tried using Aggregating Merge Tree even that didn't work well.

```
create table final_test_ag (
  CMO_ID String, 
  ROLL_UP_DATE Date, 
  CODE String, 
  value_clp AggregateFunction(uniq, Int64),
  value_plp AggregateFunction(uniq, Int64)
)
ENGINE = MergeTree
PARTITION BY toYYYYMM(ROLL_UP_DATE)
ORDER BY (CODE, CMO_ID, ROLL_UP_DATE)
SETTINGS index_granularity = 8192
```

```
SELECT CMO_ID, uniqMerge(value_clp) AS CLP, uniqMerge(value_plp) AS PLP
FROM final_test_ag
WHERE CODE IN (.......)
GROUP BY CMO_ID
LIMIT 10
```

Please suggest best method for getting approx. distinct counts

Data size 6 billion rows, Cluster has 2 Shards

**Comment by dancing-elf** [comment]
Maybe discussion in https://github.com/ClickHouse/ClickHouse/issues/18275 would be useful for you. Especially https://github.com/ClickHouse/ClickHouse/issues/18275#issuecomment-748853548

**Comment by vigneshshettyin** [comment]
Hey, @dancing-elf  I am already following what they have mentioned in the comment. My group by is wrt to CMO_ID by this each node can run query independently also I am using distributed_group_by_no_merge = 2;

With this as well the performance is up to mark.  

For now we are just upgrading hardware and checking the query performance.

**Comment by lingtaolf** [comment]
In  #18275 , they must do something while data ingestion. You must ingest the data to different servers by hashing the specified field which you need in `uniq` function. But in your scenario, you need compute 2 uniq functions. You can't hash 2 filed, because the hash of 2 fields does not guarantee that the same field value will be ingested to one shard.  
You can create 2 table , one hash by PLP_ID, another hash by CLP_ID. Then you can compute the uniq by   
```sql
SELECT 
  CMO_ID, sum(uniq_PLP_ID) 
FROM
(SELECT
    CMO_ID,
    uniq(PLP_ID) AS uniq_PLP_ID
 FROM 
    final_table
 WHERE 
    CODE IN ('LOP1', 'LOP2', 'LOP3', 'LOP4', 'LOP5', 'LOP6', 'LOP7', 'LOP8', 'LOP9', 'LOP10', 'LOP11', 'LOP12', 'LOP13', 'LOP14', 'LOP15', 'LOP16', 'LOP17', 'LOP18', 'LOP19', 'LOP20', 'LOP21', 'LOP22')
GROUP BY 
  CMO_ID
SETTINGS distributed_group_by_no_merge=1)
GROUP BY 
  CMO_ID
```  
And if you want more efficient, you can create materialized view base on above example. To number feild, I think `groupBitmap` may be more efficient.

**Reply by vigneshshettyin** [reply]
Thanks for the response!

Let me try this @lingtaolf

**Reply by lingtaolf** [reply]
For testing, you can ingest data via distributed table to hash. Eg.
```sql
 create table final_table_distrubted as  final_table Eingine = Distributed(cluster_name, db_name, final_table, cityHash64(CLP_ID));
```
Just insert data into this distributed table, it will shuffle the data to different server and same field value will be ingested into same shard.
Then query will be like:
```sql
SELECT 
  CMO_ID, sum(uniq_CLP_ID) 
FROM
(SELECT
    CMO_ID,
    uniq(CLP_ID) AS uniq_CLP_ID
 FROM 
    final_table_distributed
 WHERE 
    CODE IN ('LOP1', 'LOP2', 'LOP3', 'LOP4', 'LOP5', 'LOP6', 'LOP7', 'LOP8', 'LOP9', 'LOP10', 'LOP11', 'LOP12', 'LOP13', 'LOP14', 'LOP15', 'LOP16', 'LOP17', 'LOP18', 'LOP19', 'LOP20', 'LOP21', 'LOP22')
GROUP BY 
  CMO_ID
SETTINGS distributed_group_by_no_merge=1)
GROUP BY 
  CMO_ID
```
@vigneshshettyin

**Reply by vigneshshettyin** [reply]
Thanks for your response @lingtaolf .

Now, ORDER BY query improved by few seconds.

But still the inner query is taking time any way to optimize it for better performance.

```
SELECT
    CMO_ID,
    uniq(CLP_ID) AS uniq_CLP_ID
 FROM 
    final_table_distributed
 WHERE 
    CODE IN ('LOP1', 'LOP2', 'LOP3', 'LOP4', 'LOP5', 'LOP6', 'LOP7', 'LOP8', 'LOP9', 'LOP10', 'LOP11', 'LOP12', 'LOP13', 'LOP14', 'LOP15', 'LOP16', 'LOP17', 'LOP18', 'LOP19', 'LOP20', 'LOP21', 'LOP22')
GROUP BY 
  CMO_ID
SETTINGS distributed_group_by_no_merge=1 
```

**Reply by lingtaolf** [reply]
Have you tried it with MV.  
```sql
create materialized view final_table_mv
(
  CMO_ID String, 
  ROLL_UP_DATE Date, 
  CODE String, 
  uniqstate_CLP_ID AggregateFunction(uniq, Int64)
) Engine=AggregatingMergeTree() order by  ROLL_UP_DATE, CODE, CMO_ID as  SELECT
    CMO_ID,
    ROLL_UP_DATE,
    CODE String
    uniqState(CLP_ID) AS uniqstate_CLP_ID
 FROM 
    final_table
GROUP BY 
 -- order by low cardinality asc
  ROLL_UP_DATE, CODE, CMO_ID

create final_table_mv_distributed as final_table_mv Engine=Distributed(cluster_name, db_name, final_table_mv, rand());
```
Then ingest data into table `final_table_distributed`.
The query will be like:
```sql 
SELECT 
  CMO_ID, sum(uniq_CLP_ID) 
FROM
(SELECT
    CMO_ID,
    uniqMerge(uniqstate_CLP_ID) AS uniq_CLP_ID
 FROM 
    final_table_mv_distributed
 WHERE 
    CODE IN ('LOP1', 'LOP2', 'LOP3', 'LOP4', 'LOP5', 'LOP6', 'LOP7', 'LOP8', 'LOP9', 'LOP10', 'LOP11', 'LOP12', 'LOP13', 'LOP14', 'LOP15', 'LOP16', 'LOP17', 'LOP18', 'LOP19', 'LOP20', 'LOP21', 'LOP22')
GROUP BY 
  CMO_ID
SETTINGS distributed_group_by_no_merge=1)
GROUP BY 
  CMO_ID

```
@vigneshshettyin

**Reply by vigneshshettyin** [reply]
> Have you tried it with MV.  
> 
> ```sql
> 
> create materialized view final_table_mv
> 
> (
> 
>   CMO_ID String, 
> 
>   ROLL_UP_DATE Date, 
> 
>   CODE String, 
> 
>   uniqstate_CLP_ID AggregateFunction(uniq, Int64)
> 
> ) Engine=AggregatingMergeTree() order by  ROLL_UP_DATE, CODE, CMO_ID as  SELECT
> 
>     CMO_ID,
> 
>     ROLL_UP_DATE,
> 
>     CODE String
> 
>     uniqState(CLP_ID) AS uniqstate_CLP_ID
> 
>  FROM 
> 
>     final_table
> 
> GROUP BY 
> 
>  -- order by low cardinality asc
> 
>   ROLL_UP_DATE, CODE, CMO_ID
> 
> 
> 
> create final_table_mv_distributed as final_table_mv Engine=Distributed(cluster_name, db_name, final_table_mv, rand());
> 
> ```
> 
> Then ingest data into table `final_table_distributed`.
> 
> The query will be like:
> 
> ```sql 
> 
> SELECT 
> 
>   CMO_ID, sum(uniq_CLP_ID) 
> 
> FROM
> 
> (SELECT
> 
>     CMO_ID,
> 
>     uniqMerge(uniqstate_CLP_ID) AS uniq_CLP_ID
> 
>  FROM 
> 
>     final_table_mv_distributed
> 
>  WHERE 
> 
>     CODE IN ('LOP1', 'LOP2', 'LOP3', 'LOP4', 'LOP5', 'LOP6', 'LOP7', 'LOP8', 'LOP9', 'LOP10', 'LOP11', 'LOP12', 'LOP13', 'LOP14', 'LOP15', 'LOP16', 'LOP17', 'LOP18', 'LOP19', 'LOP20', 'LOP21', 'LOP22')
> 
> GROUP BY 
> 
>   CMO_ID
> 
> SETTINGS distributed_group_by_no_merge=1)
> 
> GROUP BY 
> 
>   CMO_ID
> 
> 
> 
> ```
> 
> @vigneshshettyin 

Yes, I tried with this as well. No improvement 

Order by ASC LOW CARDINALITY

roll up date we have data for 4 years, CMO_ID total uniq are 2.4 million and 24 unique codes.  @lingtaolf

**Reply by lingtaolf** [reply]
You need to scan all data without time range or CMO_ID ? @vigneshshettyin

**Reply by vigneshshettyin** [reply]
Yes, we have that option to get an aggregate of the entire date range. Data scanning is not a problem in our case Clickhouse handles it fast but when it comes to calculating count distinct it slows down. 

Here in the above query, if I use count instead of unique, the query gives results within 500ms.
So data scanning is not an issue 

@lingtaolf

**Reply by lingtaolf** [reply]
I konw, but if you reduce the cardinality of fields to group by , i think it will be faster. For testing, you can remove the field `ROLL_UP_DAT` from the materialized view. The number of data rows in MV will be reduced by  about 1400. If it works, You can create an extra table for this option. @vigneshshettyin

**Reply by lingtaolf** [reply]
CODE also can be removed

**Reply by lingtaolf** [reply]
If your filter include all CODE, you can remove it from MV, if not , you can remove it from group by statement in MV DDL. @vigneshshettyin

**Reply by vigneshshettyin** [reply]
Yes, that is true.

But with a year date range selected also, the query is slow.

I suppose it's a hardware issue 

Currently, our cluster has 2 shards each shard has 16vCPUs and 128 GiB Memory.

I will try removing the ROLL_UP_DATE and try. Do you think of any other solution or upgrading the hardware is the only solution?

**Reply by vigneshshettyin** [reply]
> CODE also can be removed

Yes you are right, I'll do that as well. 

Problem is this will solve for one of our use cases but CODE can be dynamically selected UI 

@lingtaolf

**Reply by lingtaolf** [reply]
You need to check the hardware statistics. If IO, CPU, Memory usage is not high, I think it's not necessary to upgrad the hardware. @vigneshshettyin

**Reply by vigneshshettyin** [reply]
> You need to check the hardware statistics. If IO, CPU, Memory usage is not high, I think it's not necessary to upgrad the hardware. @vigneshshettyin 

I observed it's almost going to 100% when the query is running 

@lingtaolf

**Reply by lingtaolf** [reply]
I konw, this is just to test and find which way can improve query. Then you can do something on your web UI. Adding limitation of code or other fields, for example, you can select a maximum of three CODEs. @vigneshshettyin

**Reply by vigneshshettyin** [reply]
> I konw, this is just to test and find which way can improve query. Then you can do something on your web UI. Adding limitation of code or other fields, for example, you can select a maximum of three CODEs. @vigneshshettyin 
> 
> 

Yes

**Reply by lingtaolf** [reply]
Ok, upgrading hardware may improve the query. If scaling resources is easy for you, you can try it.
